source/rpt/compartmentalization.cc

```diff
@@ -0,0 +1,963 @@
+#include <core/lethe_grid_tools.h>
+#include <core/solutions_output.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/grid/cell_id.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+
+#include <deal.II/lac/block_vector.h>
+
+#include <deal.II/numerics/data_component_interpretation.h>
+#include <deal.II/numerics/data_out.h>
+
+#include <bits/stdc++.h>
+#include <rpt/compartmentalization.h>
+
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+template <int dim>
+Compartmentalization<dim>::Compartmentalization(
+  CPCalculatingParameters &CPparameters)
+  : computing_timer(std::cout, TimerOutput::summary, TimerOutput::wall_times)
+  , mpi_communicator(MPI_COMM_WORLD)
+  , triangulation(mpi_communicator)
+  , cp_parameters(CPparameters)
+{}
+
+template <int dim>
+void
+Compartmentalization<dim>::generate_cylindrical_grid()
+{
+  GridOut grid_out;
+  GridGenerator::subdivided_cylinder(
+    triangulation,
+    cp_parameters.cp_param.subdivisions,
+    cp_parameters.cp_param.cylinder_radius,
+    cp_parameters.cp_param.cylinder_half_length);
+  triangulation.refine_global(cp_parameters.cp_param.initial_refinement);
+  std::ofstream output("triangulation.vtk");
+  grid_out.write_vtk(triangulation, output);
+  std::vector<double>     cell_index_output;
+  std::vector<Point<dim>> points;
+  for (const auto &cell : triangulation.active_cell_iterators())
+    {
+      // std::vector<double> center;
+      Point<3, double> center;
+      if (cell->is_locally_owned())
+        {
+          center = cell->center();
+          cells_and_indices.insert({cell->index(), cell});
+          points.push_back(center);
+          center.clear();
+        }
+    }
+
+  // output the center of cell to feed the Pyvista and extract the electric
+  // field and velocity at the cell centers
+  // from COMSOL or CFD simulation
+  std::string   filename = "cell_centers";
+  std::ofstream myfile;
+  std::string   sep = " ";
+  myfile.open(filename);
+  for (auto &point : points)
+    {
+      for (int j = 0; j < 3; j++)
+        {
+          myfile << point[j] << sep;
+        }
+      myfile << "\n";
+    }
+  myfile.close();
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::electric_field_stored_in_vector()
+  -> std::vector<std::vector<double>>
+{
+  std::vector<double> electric_field;
+  std::ifstream       fin("input_values_emw");
+  double              element;
+  while (fin >> element)
+    {
+      electric_field.push_back(element);
+    }
+
+
+  for (const auto &[cell_index, cell] : cells_and_indices)
+    {
+      std::vector<double> fill_matrix;
+      fill_matrix.push_back(cell_index);
+      fill_matrix.push_back(electric_field[cell_index]);
+      primer_matrix_index_emw.push_back(fill_matrix);
```

Comment:
If you know the number of element, it's always better to initialize it with the value and index it after. Also, you can do primer_matrix_index_emw[i] = ({cell_index, electric_field[cell_index]}); it's shorter. 
If primer_matrix_index_emw is a vector of pair instead of a vector of vector, you can you make_pair :)